Replace custom build process with WP-Scripts

iamdharmesh opened the pull request:
### Description of the Change
PR replaces current custom build process with WP-Scripts also removes some legacy files which no longer needed

Closes #193 

### How to test the Change
1. Run `npm install`
1. Run `npm run build`
1. Make sure build files are generated successfully.
1. Smoke test plugin and make sure everything working fine

### Changelog Entry
> Changed - Replace custom build process with WP-Scripts

### Credits
<!-- Please list any and all contributors on this PR so that they can be added to this projects CREDITS.md file. -->
Props @iamdharmesh @peterwilsoncc 


### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you are unsure about any of these, please ask for clarification.  We are here to help! -->
- [x] I agree to follow this project's [**Code of Conduct**](https://github.com/10up/.github/blob/trunk/CODE_OF_CONDUCT.md).
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my change.
- [ ] All new and existing tests pass.

iamdharmesh committed:
Added @wordpress/scripts removed unwanted dependencies.

iamdharmesh committed:
Updated webpack config.

iamdharmesh committed:
Update icons.

iamdharmesh committed:
Removed unwanted config.

iamdharmesh committed:
Removed external build step.

iamdharmesh committed:
Load dependencies from asset.php file.

iamdharmesh committed:
Remove legacy code.

ravinderk committed:
chore: use correct npm command

ravinderk commented on includes/admin/assets.php:
```diff
@@ -180,8 +152,18 @@ function enqueue_editor_assets() {
+			'wp-primitives',
+		),
+	);
+	if ( file_exists( $asset_file ) ) {
```
@iamdharmesh, is there any reason the asset php file will not be in the final build? I think we should remove this condition.

iamdharmesh replied:
@ravinderk If something goes wrong in the build process and the asset file is not generated then without this check plugin will throw the fatal error. I am inclined to keep this for the safe side, we are following a similar approach in our other plugins as well.

ravinderk replied:
@iamdharmesh for this we have e2e tests, right? If we do not want this to happen, then add a test for this, but do not check the file in the production code.

As I said before, there is no code issue but an approach.

cc: @dkotter

iamdharmesh replied:
> @iamdharmesh for this we have e2e tests, right?

@ravinderk Yes, we do have e2e tests. But build files are generated directly in the [deploy action](https://github.com/10up/autoshare-for-twitter/blob/develop/.github/workflows/deploy-to-wpdotorg.yml#L15) and we are not running e2e tests there before deploying the plugin. I am not a fan of adding conditions, but for the safe side, I feel it is worth it here. @dkotter any thoughts on this?

ravinderk replied:
@iamdharmesh, we run E2E tests when pushing to `develop` and `trunk`. Both are protective branches. Is this not the same thing? it make sure that `trunk` will have a valid code.

dkotter replied:
I can see both points here. In theory we shouldn't need a `file_exists` check, as this file should always be there in the final build (and if it's not there, almost certainly we have bigger problems). But as Dharmesh points out, we build the plugin right before deploy, so there's a potential scenario (however slim) that in between building the plugin to run our E2E tests on a push to `trunk` and running the same build process when tagging a release, that something may go wrong (for instance, maybe a package we use in our build process has a new release that breaks something).

Again, this is all very unlikely that our E2E tests will pass on a merge to `trunk` but then something goes wrong during the release tag process but it could happen. Because of that, I don't mind having this check as a means of defensive programming (though ideally we're also checking the final release files that make it to .org to ensure all built files exist there)

ravinderk reviewed (dismissed):
@iamdharmesh The Pull request is looking good to me. I left a minor suggestion and a new issue which is not a blocker. 

Let me know if you have any questions.


Additional issues:

- The Twitter icon is middle aligned in the column, which looks odd with other
<img width="1543" alt="image" src="https://github.com/10up/autoshare-for-twitter/assets/1784821/4fbe0fa6-e261-4519-8076-df9c4bfc6caa">

iamdharmesh committed:
Top align twitter icon.

iamdharmesh committed:
Merge branch 'enhancement/193' of github.com:10up/autoshare-for-twitter into enhancement/193

iamdharmesh committed:
TEMP: run e2e tests

iamdharmesh committed:
Switch back to develop.

peterwilsoncc commented on package.json:
```diff
@@ -40,23 +40,15 @@
-		"@wordpress/prettier-config": "^1.1.1",
-		"babel-eslint": "^10.0.3",
-		"babel-loader": "^8.0.6",
+		"@wordpress/prettier-config": "^1.4.0",
```
`@wordoress/prettier-config` and `@wordpress/eslint-plugin` are included in wp-scripts so I think they can be removed as devDependencies of this file.

`@wordpress/api-fetch` can be removed from dependencies as wp-scripts includes that as an external.

iamdharmesh replied:
Thank you for the note. This is handled in https://github.com/10up/autoshare-for-twitter/pull/282/commits/0eee6738c675de7e10e962376774968dcab388d6

peterwilsoncc commented on webpack.config.js:
```diff
@@ -1,23 +1,26 @@
+		'@wordpress/i18n': 'wp.i18n',
+		'@wordpress/plugins': 'wp.plugins',
+		lodash: 'lodash',
+		react: 'React',
```
Each of these are included in the wp-scripts externals, I'm trying to figure out where `admin-autoshare-for-twitter` is used but it can be added as a custom external if needed. See [example in Distributor](https://github.com/10up/distributor/blob/40b10f2774ba32b342396a7460e71049d0206d38/webpack.config.js#L109-L113).

iamdharmesh replied:
Fixed in https://github.com/10up/autoshare-for-twitter/pull/282/commits/ae2aae00970c19301a858ed2329451a2c62303d4

peterwilsoncc reviewed (commented):
A couple of notes inline.

iamdharmesh committed:
Remove unwanted packages.

iamdharmesh committed:
eslint fixes

jeffpaul commented:
@ravinderk @peterwilsoncc looking for updated review, thanks!

ravinderk reviewed (changes requested):
@iamdharmesh 

I noticed the following:

- Eslint list issues in the pull request.
- E2E test GitHub workflow is failing.
- Few missing text domains from translatable strings.
![image](https://github.com/10up/autoshare-for-twitter/assets/1784821/6cbdf560-d50d-48c1-8779-0104d8e77626)
![image](https://github.com/10up/autoshare-for-twitter/assets/1784821/aa2fedfc-e4ed-437b-8c83-4f394740ed49)


Let me know if you have questions.

iamdharmesh committed:
PR feedback.

iamdharmesh commented:
@ravinderk I have addressed PR feedback. Also, E2E tests are failing as we use `build-release-zip.yml` from `develop` and have made some changes in the build scripts. I temporarily pointed it to this PR branch to ensure E2E tests are running fine in this [commit](https://github.com/10up/autoshare-for-twitter/pull/282/commits/4822db6fcf7f238817f1a851bca2817be5801023) and it's E2E tests are passing. currently, I have rolled back it.
